Add more validation for auth and bib

andersju opened the pull request:
- Makes it possible for JsonLdValidator to check a set of different scopes depending on legacy collection. Figured this would be the easiest way (it's just "temporary"). Makes POST/PUT in Crud send collection type along with the data to the validator.
- For auth: enables MISSING_DEFINITION, OBJECT_PROPERTIES, UNKOWN_VOCAB_VALUE
- For bib: enables MISSING_DEFINITION, OBJECT_PROPERTIES

OBJECT_PROPERTIES corresponds to the errors OBJECT_TYPE_EXPECTED and VOCAB_STRING_EXPECTED.

jannistsiroyannis commented on whelk-core/src/main/groovy/whelk/JsonLdValidator.groovy:
```diff
@@ -41,9 +59,26 @@ class JsonLdValidator {
+            scopes.addAll(scopesPerLegacyCollection[collection])
+        }
+
+        return scopes.inject([], { errors, value ->
```
This is fine! But I had to go look it up, and it sort of hurt my head ;(
There are so many annoying ways to be clever in groovy.

andersju replied:
Was actually thinking that too. I'll make it more readable in the next iteration :)

jannistsiroyannis reviewed (approved):
:ship: good to get some more validation activated!

lrosenstrom commented:
Nice! A suggestion though would be to run doValidate once instead of once per scope. Now we traverse the graph and prevalidate everything several times per record. The scopes could be e.g. AUTH, BIB, ALL, or alternatively we could keep a list of scopes for the Validation object. Then we can decide what methods to call for a particular scope in the newly added switch/case!

andersju commented:
> Nice! A suggestion though would be to run doValidate once instead of once per scope. Now we traverse the graph and prevalidate everything several times per record. The scopes could be e.g. AUTH, BIB, ALL, or alternatively we could keep a list of scopes for the Validation object. Then we can decide what methods to call for a particular scope / set of scopes in the newly added switch statement!

Good idea, thanks! I'll cook up a new version.

andersju commented:
Simplified things and fixed wastefulness. This is assuming that we don't need "scope" to be a specific validation, but a set of validations like "all validations", "default validations", "default + collection-specific validations".
